### PR TITLE
Introduce base-compat-batteries in preparation for a 8.4-ready release

### LINKED ISSUE #36: How to deal with new things added to Prelude in the future?
A thought occurred to me when reading [this thread](https://www.reddit.com/r/haskell/comments/3t6in3/monadfail_just_landed_in_ghc_head/) about `MonadFail`. In upcoming releases of GHC, new things are going to be added to the `Prelude` which haven't existed in GHC for very long. `MonadFail`, for instance, will be added in GHC 8.8, and I believe `Semigroup` will also make it in at some point.

This poses an interesting design question for `base-compat`, because unlike other typeclasses which have been migrated to `Prelude` (e.g., `Applicative` via the AMP), `MonadFail` _et al._ are not available in `base` prior to 4.9. It seems that we'd have to do one of the following:
1. Make `base-compat` depend on `semigroups`/`fail`/whatnot so that `Semigroup`/`MonadFail`/etc. can be backported from `Prelude` properly
2. Don't give `base-compat` any new dependencies. Instead, conditionally export things from `Prelude` based on what is available from a given `base` release
3. Some sort of hybrid approach

What are people's thought on this?

### PR DESCRIPTION
**Pull request by RyanGlScott**
This introduces `base-compat-batteries`, a variant of `base-compat` that depends on compatibility packages. Lots of internal reshuffling of the repo had to occur to make this happen, so this diff is quite large. Of note:

* See `base-compat/README.md` for an exact list of things that `base-compat` compared to `base-compat-batteries`.
* The `check` test suite now uses `base-compat-batteries`, as it exposes the same API across more versions of `base` than `base-compat` does now.
* Similarly, the test suite for `base-compat` has been migrated to `base-compat-batteries`, as it makes more sense to have a single home for it, and `base-compat-batteries` is strictly more defined.

Fixes #36, and checks off the 8.4-related boxes in #24.

### CODE REVIEW

File: base-compat-batteries/src/Prelude/Compat.hs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{-# LANGUAGE CPP, NoImplicitPrelude, PackageImports #-}
+module Prelude.Compat (
+  module Base
+#if !(MIN_VERSION_base(4,9,0))
+, Semi.Semigroup((Semi.<>))
+#endif
```
Comment (phadej) [inline base-compat-batteries/src/Prelude/Compat.hs]: Would it make sense to depend and re-export `base-orphans`'s `Data.Orphans`?

File: base-compat/src/Data/Semigroup/Compat.hs [hunk 2]
```diff
@@ -0,0 +1,37 @@
+{-# LANGUAGE CPP, NoImplicitPrelude #-}
+module Data.Semigroup.Compat (
```
Comment (phadej) [inline base-compat/src/Data/Semigroup/Compat.hs]: I wonder, should there be a (haddock) note that (some) symbols are only for `base-4.9`?

File: base-compat/src/Data/Semigroup/Compat.hs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  , mtimesDefault
+  -- * Semigroups
+  , Min(..)
+  , Max(..)
+  , First(..)
+  , Last(..)
+  , WrappedMonoid(..)
+  -- * Re-exported monoids from Data.Monoid
+  , Dual(..)
+  , Endo(..)
```
Comment (phadej) [inline base-compat/src/Data/Semigroup/Compat.hs]: those could be exported for older `base` too, but it's not biggie.

File: base-compat-batteries/src/Prelude/Compat.hs (hunk 1 above)
Comment (phadej) [reply base-compat-batteries/src/Prelude/Compat.hs]: Good. Should that be documented in readme?

Comment (RyanGlScott) [reply base-compat-batteries/src/Prelude/Compat.hs]: Done.

**Pull request merged**